runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/windmill/work/refresh/DispatchedActiveWorkRefresherTest.java

```diff
@@ -0,0 +1,230 @@
+  }
+
+  private static ComputationState createComputationState(int computationIdSuffix) {
+    return createComputationState(
+        computationIdSuffix, mock(WindmillStateCache.ForComputation.class));
+  }
+
+  private static ComputationState createComputationState(
+      int computationIdSuffix, WindmillStateCache.ForComputation stateCache) {
+    return new ComputationState(
+        COMPUTATION_ID_PREFIX + computationIdSuffix,
+        new MapTask().setStageName("stageName").setSystemName("systemName"),
+        workExecutor(),
+        new HashMap<>(),
+        stateCache);
+  }
+
+  private ActiveWorkRefresher createActiveWorkRefresher(
+      Supplier<Instant> clock,
+      int activeWorkRefreshPeriodMillis,
+      int stuckCommitDurationMillis,
+      Supplier<Collection<ComputationState>> computations,
+      Consumer<Map<String, List<HeartbeatRequest>>> activeWorkRefresherFn) {
+    return new DispatchedActiveWorkRefresher(
+        clock,
+        activeWorkRefreshPeriodMillis,
+        stuckCommitDurationMillis,
+        computations,
+        DataflowExecutionStateSampler.instance(),
+        activeWorkRefresherFn,
+        Executors.newSingleThreadScheduledExecutor());
+  }
+
+  private Work createOldWork(int workIds, Consumer<Work> processWork) {
+    return Work.create(
+        Windmill.WorkItem.newBuilder()
+            .setWorkToken(workIds)
+            .setCacheToken(workIds)
+            .setKey(ByteString.EMPTY)
+            .setShardingKey(workIds)
+            .build(),
+        DispatchedActiveWorkRefresherTest.A_LONG_TIME_AGO,
+        ImmutableList.of(),
+        processWork);
+  }
+
+  @Test
+  public void testActiveWorkRefresh() {
+    int activeWorkRefreshPeriodMillis = 100;
+
+    // Block work processing to queue up the work.
+    CountDownLatch workIsProcessed = new CountDownLatch(1);
+    Consumer<Work> processWork =
+        ignored -> {
+          try {
+            workIsProcessed.await();
+          } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+          }
+        };
+
+    List<ComputationState> computations = new ArrayList<>();
+    Map<String, List<Work>> computationsAndWork = new HashMap<>();
+    for (int i = 0; i < 5; i++) {
+      ComputationState computationState = createComputationState(i);
+      Work fakeWork = createOldWork(i, processWork);
+      computationState.activateWork(ShardedKey.create(ByteString.EMPTY, i), fakeWork);
+
+      computations.add(computationState);
+      List<Work> activeWorkForComputation =
+          computationsAndWork.computeIfAbsent(
+              computationState.getComputationId(), ignored -> new ArrayList<>());
+      activeWorkForComputation.add(fakeWork);
+    }
+
+    Map<String, List<HeartbeatRequest>> expectedHeartbeats = new HashMap<>();
+
+    ActiveWorkRefresher activeWorkRefresher =
+        createActiveWorkRefresher(
+            Instant::now,
```

Comment:
Can we send a fake clock and control the passage of time instead of relying on sleeping?